vernac/obligations.ml

```diff
@@ -827,17 +827,24 @@ let rec string_of_list sep f = function
 
 (* Solve an obligation using tactics, return the corresponding proof term *)
 
-let solve_by_tac name evi t poly ctx =
-  let id = name in
-  (* spiwack: the status is dropped. *)
-  let (entry,_,ctx') = Pfedit.build_constant_by_tactic
-    id ~goal_kind:(goal_kind poly) ctx evi.evar_hyps evi.evar_concl (Tacticals.New.tclCOMPLETE t) in
-  let env = Global.env () in
-  let entry = Safe_typing.inline_private_constants_in_definition_entry env entry in
-  let body, () = Future.force entry.const_entry_body in
-  let ctx' = Evd.merge_context_set ~sideff:true Evd.univ_rigid (Evd.from_ctx ctx') (snd body) in
-  Inductiveops.control_only_guard env ctx' (EConstr.of_constr (fst body));
-  (fst body), entry.const_entry_type, Evd.evar_universe_context ctx'
+let solve_by_tac loc name evi t poly ctx =
+  try
+    let id = name in
+    (* spiwack: the status is dropped. *)
+    let (entry,_,ctx') = Pfedit.build_constant_by_tactic
+        id ~goal_kind:(goal_kind poly) ctx evi.evar_hyps evi.evar_concl (Tacticals.New.tclCOMPLETE t) in
+    let env = Global.env () in
+    let entry = Safe_typing.inline_private_constants_in_definition_entry env entry in
+    let body, () = Future.force entry.const_entry_body in
+    let ctx' = Evd.merge_context_set ~sideff:true Evd.univ_rigid (Evd.from_ctx ctx') (snd body) in
+    Inductiveops.control_only_guard env ctx' (EConstr.of_constr (fst body));
+    Some ((fst body), entry.const_entry_type, Evd.evar_universe_context ctx')
+  with e when CErrors.noncritical e ->
+    let (e, _) = CErrors.push e in
+    match e with
+    | Refiner.FailError (_, s) ->
+      user_err ?loc ~hdr:"solve_obligation" (Lazy.force s)
+    | e -> None (* FIXME really ? *)
```

Comment:
I guess the refactoring to option is fine, however I think you should strengthen the semantics here [CI permitting] and either:

- turn the `tclCOMPLETE` into a real exception.
- or even better remote `tclCOMPLETE` at all, it should be close_proof the one in charge of owning this logic. Let me have indeed a go at it.